src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -511,27 +376,97 @@ public void run() {
      * @return A Completable that completes when the connection has been stopped.
      */
     private Completable stop(String errorMessage) {
-        hubConnectionStateLock.lock();
+        Transport transport;
+        this.state.lock();
         try {
-            if (hubConnectionState == HubConnectionState.DISCONNECTED) {
+            if (this.state.getHubConnectionState() == HubConnectionState.DISCONNECTED) {
                 return Completable.complete();
             }
 
             if (errorMessage != null) {
-                stopError = errorMessage;
+                this.state.getConnectionStateUnsynchronized().stopError = errorMessage;
                 logger.error("HubConnection disconnected with an error: {}.", errorMessage);
             } else {
                 logger.debug("Stopping HubConnection.");
             }
+
+            transport = this.state.getConnectionStateUnsynchronized().transport;
         } finally {
-            hubConnectionStateLock.unlock();
+            this.state.unlock();
         }
 
         Completable stop = transport.stop();
         stop.onErrorComplete().subscribe();
         return stop;
     }
 
+    private void ReceiveLoop(ByteBuffer payload)
+    {
+        ConnectionState connectionState = this.state.getConnectionState();
```

Comment:
This is called directly by the transport without a lock right? I know connectionState.tryRemoveInvocation and getInvocation do the necessary locking internally, but I think it might be a good idea to lock until the for loop starts to be safe.

Reply:
this locks and throws if the connection isn't active

Reply:
That lock is only take long enough to retrieve a valid connectionState instance. I am suggesting holding the lock until just before the for loop starts so we don't have to worry about resetServerTimeout, handleHandshake or whatever protocol.parseMessages is doing to the connectionState messing with other code using connectionState inside a lock. Unless there's some reason we cannot lock for longer, I don't see why we wouldn't to be safe.

I'd be for keeping the lock for the entire ReceiveLoop method if it wasn't calling user callbacks inline.